### PR TITLE
Make `UnderlineInputBorder` consistent

### PR DESCRIPTION
**Pull request by bernaferrari**
This was easy to implement. I like the result, I think `borderRadius.zero` -> `borderRadius.circular` makes a nice transition, and many places (like macOS) use an effect similar to this PR, while Google doesn't use anywhere (yet). I'm curious if it is going to break goldens or google testing.

ttt

What do you think? cc @HansMuller @gspencergoog. Is this something you want, should I ask the community, or do you prefer the current one?

Side effects:
- This makes strokeAlign work with `UnderlineInputBorder` (TODO: fix `drawLine` when borderRadius is zero).
- This is faster than the current implementation (clip is slow on Skia). 🚀 
- We could just call `BoxBorder._paintNonUniformBorder` (if it weren't private). Single LOC implementation.
- Web does this by default:
image
- Apparently no tests fail and most usages around (via code search) seem to be without a borderRadius.

### CODE REVIEW

File: packages/flutter/lib/src/material/input_border.dart
Comment (justinmc) [inline packages/flutter/lib/src/material/input_border.dart:270]: Why not use [clampDouble](https://api.flutter.dev/flutter/package-material_color_utilities_utils_math_utils/MathUtils/clampDouble.html) on these?

**Review by justinmc — COMMENTED** [review]
@bernaferrari Weren't you and I discussing this on another issue somewhere?  Could you link that if you still have the link? And post the code to produce these fields.

Also, can you clarify what the difference is between "Currently" and "Proposal" in your image?  Should the top one say BorderRadius.zero?

I couldn't get the Google tests to show up here but I will run them locally.

**Comment by justinmc** [comment]
I don't see any Google test failures.

File: packages/flutter/lib/src/material/input_border.dart
Comment (bernaferrari) [reply packages/flutter/lib/src/material/input_border.dart:270]: Because there are two values, so the other possible way would be

Radius(clampDouble(min: 0, rect.tlRadiusX + insets.left), clampDouble(min: 0, rect.tlRadiusY + inset.top))

Do you prefer that way? I can make it.

Comment (bernaferrari) [reply packages/flutter/lib/src/material/input_border.dart:270]: Oh, good news: I copied the code from what I'm doing on BoxDecoration. When my other PR gets merged, this will be redundant and I'll be able to use the `BoxDecoration.paintNonUniformBorder`.

Comment (justinmc) [reply packages/flutter/lib/src/material/input_border.dart:270]: I guess the clampDouble way doesn't look much worse, I say use that.  The linter rule is there because `.clamp` is much slower, see https://github.com/flutter/flutter/issues/103917.

File: packages/flutter/lib/src/material/input_border.dart
Comment (justinmc) [inline packages/flutter/lib/src/material/input_border.dart:255]: Nit: Could this be static?

**Review by justinmc — APPROVED** [review]
LGTM :+1: 

Thanks for the Discord link, that's what I was thinking of.  I'm on board with "Proposal" approach.

Leaving this with no autosubmit for now as requested.

File: packages/flutter/lib/src/material/input_border.dart
Comment (bernaferrari) [reply packages/flutter/lib/src/material/input_border.dart:255]: Yes (but why?).

But this is going to probably be replaced by BoxBorder.paintNonUniformPath. I just need to get that approved. Shouldn't take more than a few days.

**Comment by Hixie** [comment]
Spoke with @bernaferrari, he says this is blocked on https://github.com/flutter/flutter/pull/124417

**Comment by bernaferrari** [comment]
Yes, thanks! I forgot to write this here lol

File: packages/flutter/lib/src/material/input_border.dart
Comment (bernaferrari) [inline packages/flutter/lib/src/material/input_border.dart:253]: @justinmc Ok, heeeere we go! Not sure I need another review (probably not), just double checking. I implemented what I wanted to, and the only slightly "odd" thing is this.
Reason: If topLeft/topRight borders are there, the paint causes weird artifacts to show, so making them square, and limiting the bottomLeft/Right to h/2 prevents the issue from ever appearing.
Consequences: theoretically an elliptical border(999, 0) wouldn't be possible anymore (clamp to h/2), but a) it is not breaking, b) it is already rare in standard widgets, c) this would be even more rare. I think it is ok.

**Review by Hixie — APPROVED** [review]
LGTM

**Comment by Hixie** [comment]
@bernaferrari Ok let's try landing it!

**Pull request merged**
